Avoid invalidating SVG resources when referencing element changes style

heycam opened the pull request:
#### fda943aadc10f120d2b316f616372209eeeae430
<pre>
Avoid invalidating SVG resources when referencing element changes style
<a href="https://bugs.webkit.org/show_bug.cgi?id=242425">https://bugs.webkit.org/show_bug.cgi?id=242425</a>
&lt;rdar://problem/96572561&gt;

Reviewed by Nikolas Zimmermann.

SVGResourcesCache::clientStyleChanged will destroy then recreate the SVG
resources for an element when it is restyled and the StyleDifference is not
Repaint or RepaintIfText. This is not necessary for many property changes. We
only need to do this if there is a change in which resource is being referenced
(i.e., the clip-path, mask-image, fill, etc. property changed), or if there is
a property change on the element that a cached resource depends on.

For now, we don&apos;t bother to pinpoint the specific resources that need
removing, and instead continue to destroy and recreate the entire
SVGResources object when a relevant property changes. Doing this in a
more targeted way would not be trivial, since SVGResources::buildCachedResources
handles cyclic dependency removal.

* Source/WebCore/rendering/svg/LegacyRenderSVGModelObject.cpp:
(WebCore::LegacyRenderSVGModelObject::styleDidChange):
* Source/WebCore/rendering/svg/LegacyRenderSVGRoot.cpp:
(WebCore::LegacyRenderSVGRoot::layout):
(WebCore::LegacyRenderSVGRoot::styleDidChange):
* Source/WebCore/rendering/svg/RenderSVGBlock.cpp:
(WebCore::RenderSVGBlock::styleDidChange):
* Source/WebCore/rendering/svg/RenderSVGInline.cpp:
(WebCore::RenderSVGInline::styleDidChange):
* Source/WebCore/rendering/svg/RenderSVGModelObject.cpp:
(WebCore::RenderSVGModelObject::styleDidChange):
* Source/WebCore/rendering/svg/RenderSVGResourceContainer.cpp:
(WebCore::RenderSVGResourceContainer::registerResource):
* Source/WebCore/rendering/svg/RenderSVGRoot.cpp:
(WebCore::RenderSVGRoot::layout):
(WebCore::RenderSVGRoot::styleDidChange):
* Source/WebCore/rendering/svg/SVGResourcesCache.cpp:
(WebCore::SVGResourcesCache::clientStyleChanged):
* Source/WebCore/rendering/svg/SVGResourcesCache.h:

Canonical link: <a href="https://commits.webkit.org/252484@main">https://commits.webkit.org/252484@main</a>
</pre>

nikolaszimmermann commented on Source/WebCore/rendering/svg/RenderSVGRoot.cpp:
```diff
@@ -385,7 +385,7 @@ void RenderSVGRoot::willBeRemovedFromTree(IsInternalMove isInternalMove)
 {
     RenderReplaced::styleDidChange(diff, oldStyle);
-    SVGResourcesCache::clientStyleChanged(*this, diff, style());
+    SVGResourcesCache::clientStyleChanged(*this, diff, oldStyle, style());
```
I'm happy the day the SVGResourcesCache goes away - early design mistakes, to use renderers to manage resource lookup/caching/... :/

nikolaszimmermann commented on Source/WebCore/rendering/svg/SVGResourcesCache.cpp:
```diff
@@ -96,7 +96,41 @@ static inline bool rendererCanHaveResources(RenderObject& renderer)
 }
 
-void SVGResourcesCache::clientStyleChanged(RenderElement& renderer, StyleDifference diff, const RenderStyle& newStyle)
+static bool hasStyleDifferencesAffectingResources(RenderObject& renderer, const RenderStyle* oldStyle, const RenderStyle& newStyle)
```
Any specific reason for a static function rather than a lambda, at the point where it's used?

heycam replied:
No, I can change it to a lambda.

nikolaszimmermann reviewed (commented):
Thumbs up for this change! Long overdue to cleanup RenderSVGResource & co.
LBSE will also bring some changes in that area in future (LBSE downstream no longer stores any state for RenderSVGResourceClipper/Filter/Masker, and the actual clipping/filter/masking logic is (re-)integrated within the RenderLayer based clipping / masking code (and RenderLayerFilters fully supports the remaining necessary SVG features). Upstreaming that is still a while away though... Just wanted to let you know about this, since you've touched that area.

I have a general comment/question: Are you aware that we ~ O(200) tests in svg/ that rely solely on pixel tests to catch painting regressions? This goes back to at least 2007 where we still attempted to run pixel tests across EWS & co. Unfortunately many repaint/invalidation/... tests have not been converted to ref-tests yet -- therefore we have quite a large surface, not covered by regular testing.

Given the nature of this change (optimize from always-invalidate to selective-invalidate).. I'd strongly suggest you to generate yourself a pixel test baseline for the whole svg/ subdirectory (rwt --pixel-tests --new-baseline svg/) using a vanilla build, and afterwards run the pixel tests with your change applied (rwt --pixel-tests --tolerance=0 svg/). Would be interesting to see if you spot any changes, not found by regular testing. These would be good candidates for new reftests :-)

nikolaszimmermann commented on Source/WebCore/rendering/svg/SVGResourcesCache.cpp:
```diff
@@ -105,11 +105,42 @@ void SVGResourcesCache::clientStyleChanged(RenderElement& renderer, StyleDiffere
+        if (oldStyle->filter() != newStyle.filter())
+            return true;
+
+        // -apple-color-filter affects gradients.
```
I wasn't aware that this can affect SVG? Ideally we'd need tests for each of the if statements here. Is that done using the other bugs you filed, I guess not for apple-color-filter at least (not really sure if we need to support that..).

heycam replied:
Right, the tests in the other bug are meant to cover these here. But I don't have one for `-apple-color-filter`. I'll add one.

heycam replied:
I've added a sub-test for this in #2213.

nikolaszimmermann commented on Source/WebCore/rendering/svg/SVGResourcesCache.cpp:
```diff
@@ -105,11 +105,42 @@ void SVGResourcesCache::clientStyleChanged(RenderElement& renderer, StyleDiffere
-        auto& cache = resourcesCacheFromRenderer(renderer);
+    auto& cache = resourcesCacheFromRenderer(renderer);
+
+    auto hasStyleDifferencesAffectingResources = [&] {
```
I thought [&]() is preferred?

heycam replied:
For the type of the variable holding the lambda? I don't think so, at least not to store the just-created lambda.

```
$ rg '^ *auto \S+ = \[&\]' Source/ | wc -l
    1700
$ rg '^ *auto& \S+ = \[&\]' Source/ | wc -l
       4
```

nikolaszimmermann replied:
> For the type of the variable holding the lambda? I don't think so, at least not to store the just-created lambda.

Lol, sorry for that, you need to view the source of my comment to understand it ;-)
```I thought [&]() is preferred?```

I was referring to the pair of braces afterwards.

heycam replied:
Aha, `[&]()` was interpreted as a Markdown link. :-)

```
$ rg '\[&\] \{' Source/ | wc -l
    1266
$ rg '\[&\] ?\(\) ?\{' Source/ | wc -l
     397
```

And [looks like no parens is preferred](https://webkit.org/code-style-guidelines/#punctuation-omit-lambda-paren).

nikolaszimmermann reviewed (commented):
Almost there, r=me, however some questions remain. I'll approve and set merging-blocked until you had a chance to comment (is that the right GitHub workflow? r=me+comments == r+ + merging-blocked label?)

nikolaszimmermann reviewed (approved):
LGTM.

heycam committed:
Avoid invalidating SVG resources when referencing element changes style
https://bugs.webkit.org/show_bug.cgi?id=242425
<rdar://problem/96572561>

Reviewed by Nikolas Zimmermann.

SVGResourcesCache::clientStyleChanged will destroy then recreate the SVG
resources for an element when it is restyled and the StyleDifference is not
Repaint or RepaintIfText. This is not necessary for many property changes. We
only need to do this if there is a change in which resource is being referenced
(i.e., the clip-path, mask-image, fill, etc. property changed), or if there is
a property change on the element that a cached resource depends on.

For now, we don't bother to pinpoint the specific resources that need
removing, and instead continue to destroy and recreate the entire
SVGResources object when a relevant property changes. Doing this in a
more targeted way would not be trivial, since SVGResources::buildCachedResources
handles cyclic dependency removal.

* Source/WebCore/rendering/svg/LegacyRenderSVGModelObject.cpp:
(WebCore::LegacyRenderSVGModelObject::styleDidChange):
* Source/WebCore/rendering/svg/LegacyRenderSVGRoot.cpp:
(WebCore::LegacyRenderSVGRoot::layout):
(WebCore::LegacyRenderSVGRoot::styleDidChange):
* Source/WebCore/rendering/svg/RenderSVGBlock.cpp:
(WebCore::RenderSVGBlock::styleDidChange):
* Source/WebCore/rendering/svg/RenderSVGInline.cpp:
(WebCore::RenderSVGInline::styleDidChange):
* Source/WebCore/rendering/svg/RenderSVGModelObject.cpp:
(WebCore::RenderSVGModelObject::styleDidChange):
* Source/WebCore/rendering/svg/RenderSVGResourceContainer.cpp:
(WebCore::RenderSVGResourceContainer::registerResource):
* Source/WebCore/rendering/svg/RenderSVGRoot.cpp:
(WebCore::RenderSVGRoot::layout):
(WebCore::RenderSVGRoot::styleDidChange):
* Source/WebCore/rendering/svg/SVGResourcesCache.cpp:
(WebCore::SVGResourcesCache::clientStyleChanged):
* Source/WebCore/rendering/svg/SVGResourcesCache.h:

Canonical link: https://commits.webkit.org/252484@main
